### ISSUE TITLE
win_update: "Failed to search for updates: Exception from HRESULT: 0x80244007"

### ISSUE THREAD

**Issue by Udayendu** [ansible-collections/ansible.windows#274]
##### SUMMARY
**win_update** module is not able to search the update and showing the following error message:

```
FAILED! => {"changed": false, "filtered_updates": {}, "msg": "Failed to search for updates: Exception from HRESULT: 0x80244007", "updates": {}}
```

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
- win_update

##### ANSIBLE VERSION
```
$ pip3 show ansible-core
Name: ansible-core
Version: 2.11.1
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3+
Location: /usr/local/lib/python3.8/dist-packages
Requires: cryptography, resolvelib, jinja2, packaging, PyYAML
Required-by: ansible

$ pip3 show ansible
Name: ansible
Version: 4.1.0
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3+
Location: /usr/local/lib/python3.8/dist-packages
Requires: ansible-core

$ pip3 show pywinrm
Name: pywinrm
Version: 0.4.2
Summary: Python library for Windows Remote Management
Home-page: http://github.com/diyan/pywinrm/
Author: Alexey Diyan
Author-email: alexey.diyan@gmail.com
License: MIT license
Location: /usr/local/lib/python3.8/dist-packages
Requires: requests, six, xmltodict, requests-ntlm
Required-by:
```

##### COLLECTION VERSION
```
$ ansible-galaxy collection list ansible.windows

# /usr/local/lib/python3.8/dist-packages/ansible_collections
Collection      Version
--------------- -------
ansible.windows 1.6.0
```

##### CONFIGURATION
- No specific configuration

##### OS / ENVIRONMENT
- Ansible Provision Server: 
    - Ubuntu 20.04 LTS
- Target OS: 
    - Windows Server 2016
    - Windows Server 2019

##### STEPS TO REPRODUCE
- Use the below playbook as an ansible role:
```
---
- name: Wait for system to become reachable over WinRM
  wait_for_connection:
    sleep: 5
    delay: 30
    connect_timeout: 120
    timeout: 900
  register: result

- debug: var=result

- name: Install all security, critical, and rollup updates
  win_updates:
    category_names:
      - Updates
      - SecurityUpdates
      - CriticalUpdates
      - UpdateRollups
    reboot: yes
    state: installed
    reboot_timeout: 3600
  register: result

# Output of installed patches
- debug: var=result

- name: WinRM Port Status
  wait_for:
    port: 5986
    state: started
    delay: 10
    host: "{{ inventory_hostname }}"
    timeout: 600
  register: result
  delegate_to: localhost

- debug: var=result
```

##### EXPECTED RESULTS
- It should work properly without any error message.

##### ACTUAL RESULTS
- Ansible failed to update using the following error:
```
FAILED! => {"changed": false, "filtered_updates": {}, "msg": "Failed to search for updates: Exception from HRESULT: 0x80244007", "updates": {}}
```

- But when running the update manually with in the OS, its working well.

**Comment by jborean93** [comment]
There's a tiny chance the latest release (`1.7.x`) of this collection will fix your issue so I recommend you try that out. Even if it doesn't it will at least give you a better error message. The error `0x80244007` refers to [WU_E_PT_SOAPCLIENT_SOAPFAULT](https://github.com/ansible-collections/ansible.windows/blob/2ab22bd6031778f49ddb2b04921062daa715bd0a/plugins/action/win_updates.py#L259-L260). MS gives some more info on this error https://docs.microsoft.com/en-us/windows/deployment/update/windows-update-errors

> This issue occurs because Windows cannot renew the cookies for Windows Update.  Review [KB2883975](https://support.microsoft.com/en-us/topic/0x80244007-error-when-windows-tries-to-scan-for-updates-on-a-wsus-server-6af342d9-9af6-f3bb-b6ad-2be56bf7826e) for instructions to resolve the issue.

It doesn't seem to apply to the OS you've mentioned though so I doubt the KB will solve the problem. Is your environment set up with a WSUS server or is it connecting directly to Microsoft Update? https://superuser.com/questions/810122/windows-update-failing-with-error-80244007-could-not-search-for-new-updates/825471 gives 3 different reasons why this has appeared

* Improper WSUS configuration
* `Datastore.edb` is corrupted
* Time not set up properly

Unfortunately there's very little help on the Windows Update documentation as to why this might occur. I recommend you run [Get-WindowsUpdateLog](https://docs.microsoft.com/en-us/powershell/module/windowsupdate/get-windowsupdatelog?view=windowsserver2019-ps) after a failed run to see if there is any more info in the log file as to why it's failing. From Ansible's perspective we are calling the COM API to search/download/install updates and are at the mercy of what it returns.

**Comment by Udayendu** [comment]
@jborean93 
Thanks for your response.

- Is your environment set up with a WSUS server or is it connecting directly to Microsoft Update?
   -> Yes, my env is mapped to WSUS server. No direct access to internet.

- Improper WSUS configuration
   -> Not really. Because once the update failed via ansible, if I will login to the OS and start the update, its working well.

- Datastore.edb is corrupted
   -> May or may not be. I don't know how to check. Any pointers would be highly appreciated. 

- Time not set up properly
   -> Time is correctly set based on the deployment region and its configured via NTP servers. Have checked that and time is always showing correct with in the OS level.

Also few more observation:
- If I will try to run the update 10 times, after 4-5 iteration it will work.
- Yesterday before reporting this issue, I deployed 4 vms. On first attempt, update failed on all the 4 vms. On second attempt, update worked on 2 and failed on remaining 2. On 3rd, again failed on the remaining 2 and on 4th attempt, it worked on the remaining two.

**Note:** These systems are part of the domain.

I have upgraded the ansible collections. I will run the test and update soon.

**Comment by jborean93** [comment]
> May or may not be. I don't know how to check. Any pointers would be highly appreciated.

I would check the link I shared https://superuser.com/questions/810122/windows-update-failing-with-error-80244007-could-not-search-for-new-updates/825471 as that had some more info on this problem.

Unfortunately with these problems it's hard to diagnose what's wrong because we cannot replicate the issue. It's almost certainly an environment specific setup that causes the problem. It working on an interactive session is puzzling but it wouldn't be the first time this has happened. The `win_updates` code runs in a scheduled task (older versions used become as `SYSTEM`) so the user context behind the code is slightly different. Using the updated collection version might yield positive results as it does away with the become as `SYSTEM` method which caused some unknown failures in the past and who knows it might fix this problem for you.

This is what I recommend you try the [Get-WindowsUpdateLog](https://docs.microsoft.com/en-us/powershell/module/windowsupdate/get-windowsupdatelog?view=windowsserver2019-ps) cmdlet to try and get some more info. It might indicate what exactly has gone wrong when running through Ansible. You can always compare and contrast it.

Finally you can try and run this code manually on an interactive logon and see if it also suffers from the same problem. It's essentially the same code that Ansible is running, just in a scheduled task.

```powershell
$session = New-Object -ComObject Microsoft.Update.Session
$searcher = $session.CreateUpdateSearcher()
$searcher.ServerSelection = 0
$searchResult = $searcher.Search("IsInstalled = 0")
```

**Comment by Udayendu** [comment]
@jborean93 
Thanks for the update. I have upgraded the ansible collections to the latest 1.7.2 but still its failing.
Going to do some analysis from OS side and WSUS server side and then get back to you with my findings.

**Comment by Udayendu** [comment]
In the windows update logs got no clue:

**Comment by Udayendu** [comment]
@jborean93 
I did some change in the playbook and included:
- search option before install
- and install with multiple re-try

and I can see that its working. I will do more testing and let you know how its going. Installed option supposed to do the search & install but seems like its not doing the search really well.

**Comment by jborean93** [comment]
The `No Format Information found` is a result of the host not being able to reach the Microsoft symbol server https://serverfault.com/questions/985900/get-windowsupdatelog-no-format-information-found and https://maartendamen.com/get-windowsupdatelog-no-format-information-found/. Have a look at those links to see a way of getting the logs to an internet connected host so you can run `Get-WindowsUpdateLog` properly.

Have you tried to run that PowerShell code I shared in https://github.com/ansible-collections/ansible.windows/issues/274#issuecomment-919493824 manually on the affected hosts? This is essentially what the `win_updates` module does to kick off a search and where it's failing for you. As you can see it's quite simple so I'm unsure what else we can do to fix the problem.

Ultimately there's some sort of communication issue with the client when running through the COM API and the WSUS server. All this communication is done in the black box that is the Windows Update agent and we have no idea what is actually happening. The closest we can get to finding out more information is through `Get-WindowsUpdateLog`.

**Comment by Udayendu** [comment]
> 
> 
> The `No Format Information found` is a result of the host not being able to reach the Microsoft symbol server https://serverfault.com/questions/985900/get-windowsupdatelog-no-format-information-found and https://maartendamen.com/get-windowsupdatelog-no-format-information-found/. Have a look at those links to see a way of getting the logs to an internet connected host so you can run `Get-WindowsUpdateLog` properly.
> 
> Have you tried to run that PowerShell code I shared in [#274 (comment)](https://github.com/ansible-collections/ansible.windows/issues/274#issuecomment-919493824) manually on the affected hosts? This is essentially what the `win_updates` module does to kick off a search and where it's failing for you. As you can see it's quite simple so I'm unsure what else we can do to fix the problem.

       Yes, I tried that but didn't received any output. But no error though.

> 
> Ultimately there's some sort of communication issue with the client when running through the COM API and the WSUS server. All this communication is done in the black box that is the Windows Update agent and we have no idea what is actually happening. The closest we can get to finding out more information is through `Get-WindowsUpdateLog`.

       After including search option before install, I can see that its working. But if I am directly trying to use the installed, its not able to search the packages always. 

-  So for now seems like I have a working solution but question is:
    -  if search + install is working then why not install directly ?

**Comment by jborean93** [comment]
> if search + install is working then why not install directly ?

I don't fully understand your question here. The `win_updates` module always runs the search function regardless of what `state` is set to. It needs to search for updates for it to then subsequently download/install.

**Comment by Udayendu** [comment]
> > if search + install is working then why not install directly ?
> 
> I don't fully understand your question here. The `win_updates` module always runs the search function regardless of what `state` is set to. It needs to search for updates for it to then subsequently download/install.

Well, if tha'ts how the design is then its not working as per the design. Its not capable of searching the updates properly. But if I will initiate the search using a different task and then using another task to install, its working well.

**Comment by jborean93** [comment]
What I'm trying to say is that doing `state: searched` and then `state: installed` has no difference at the point it is failing for you. The code run to get to that point in both scenarios is exactly the same and the differences only come in after that step. I cannot fathom how it would work in one of those scenarios and not the other, the only explanation is you are getting lucky and it's failing due to some external problem that isn't present at all times.

**Comment by Udayendu** [comment]
> 
> 
> What I'm trying to say is that doing `state: searched` and then `state: installed` has no difference at the point it is failing for you. The code run to get to that point in both scenarios is exactly the same and the differences only come in after that step. I cannot fathom how it would work in one of those scenarios and not the other, the only explanation is you are getting lucky and it's failing due to some external problem that isn't present at all times.

I am not sure why its making a difference but in my case first search and then install is working well. But direct install is working 20% time and failing 80% time.

Here is the code that I am using:

```
---
- name: Wait for system to become reachable over WinRM
  wait_for_connection:
    sleep: 5
    delay: 30
    connect_timeout: 120
    timeout: 900
  register: result

- debug: var=result


- name: Search all security, critical, and rollup updates
  win_updates:
    category_names:
      - Updates
      - SecurityUpdates
      - CriticalUpdates
      - UpdateRollups
    state: searched
  register: result
  until: result is not failed
  retries: 100
  delay: 6

# Output of searched patches
- debug: var=result

- name: Install all security, critical, and rollup updates
  win_updates:
    category_names:
      - Updates
      - SecurityUpdates
      - CriticalUpdates
      - UpdateRollups
    reboot: yes
    state: installed
    reboot_timeout: 3600
  register: result
  until: result is not failed
  retries: 100
  delay: 6

# Output of installed patches
- debug: var=result
```

**Comment by Udayendu** [comment]
@jborean93 
If you feel that currently nothing much can be done and module is working as designed, then we can close it.
Because the new logic is working well for me.

Thanks a lot for your help so far.
